packages/babel-traverse/src/scope/index.js

```diff
@@ -181,20 +162,19 @@ export default class Scope {
    * within.
    */
 
-  constructor(path: NodePath, parentScope?: Scope) {
-    if (parentScope && parentScope.block === path.node) {
-      return parentScope;
+  constructor(path: NodePath) {
+    const { node } = path;
+    const cached = scopeCache.get(node);
+    // Sometimes, a scopable path is placed higher in the AST tree.
+    // In these cases, have to create a new Scope.
+    if (cached && cached.path === path) {
+      return cached;
     }
-
-    const cached = getCache(path, parentScope, this);
-    if (cached) return cached;
+    scopeCache.set(node, this);
 
     this.uid = uid++;
-    this.parent = parentScope;
-    this.hub = path.hub;
 
-    this.parentBlock = path.parent;
-    this.block = path.node;
```

Comment:
Scope is also exposed to Plugins? Then removing props is a breaking change I guess.
Should land it before 7 then and add to migration guide :)